File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.java
//{ Driver Code Starts
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int testCases = Integer.parseInt(br.readLine());

        while (testCases-- > 0) {
            // Read capacity
            int capacity = Integer.parseInt(br.readLine());

            // Read values
            String[] valInput = br.readLine().split(" ");
            int[] val = new int[valInput.length];
            for (int i = 0; i < valInput.length; i++) {
                val[i] = Integer.parseInt(valInput[i]);
            }

            // Read weights
            String[] wtInput = br.readLine().split(" ");
            int[] wt = new int[wtInput.length];
            for (int i = 0; i < wtInput.length; i++) {
                wt[i] = Integer.parseInt(wtInput[i]);
            }

            // Call the knapSack function and print the result
            System.out.println(Solution.knapSack(capacity, val, wt));
            System.out.println("~");
        }
    }
}

// } Driver Code Ends


class Solution {
    // Function to return max value that can be put in knapsack of capacity.
    static int dp[][] = new int[1001][1001];
    static int knapSack(int capacity, int val[], int wt[]) {
        // code here
        int n = wt.length;
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= capacity; j++) {
                dp[i][j] = -1;
            }
        }
        return knapSackHelper(capacity,val,wt,wt.length);
    }
    static int knapSackHelper(int capacity, int val[] , int wt[], int n){
        if (capacity == 0 || n == 0) {
        dp[n][capacity] = 0;
        return 0;
    }       
        if(dp[n][capacity]!=-1)
        return dp[n][capacity];
        if(wt[n-1]<=capacity){
            dp[n][capacity] = Math.max(val[n-1]+knapSackHelper(capacity-wt[n-1],val,wt,n-1),knapSackHelper(capacity,val,wt,n-1));
        }
        else{
            dp[n][capacity] = knapSackHelper(capacity,val,wt,n-1); 
        }
        return dp[n][capacity];
    }
}

File: Difficulty: Medium/0 - 1 Knapsack Problem/0-1-knapsack-problem.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
  int dp[1001][1001];
    // Function to return max value that can be put in knapsack of capacity.
    int knapSack(int capacity, vector<int> &val, vector<int> &wt) {
        // code here
        int n = wt.size();
        memset(dp,-1,sizeof(dp));
        return knapSackHelper(capacity,val,wt,n);
    }
    int knapSackHelper(int capacity, vector<int> &val , vector<int> &wt, int n){
        if (capacity == 0 || n == 0) {
        return 0;
    }       
        if(dp[n][capacity]!=-1)
        return dp[n][capacity];
        if(wt[n-1]<=capacity){
            dp[n][capacity] = max(val[n-1]+knapSackHelper(capacity-wt[n-1],val,wt,n-1),knapSackHelper(capacity,val,wt,n-1));
        }
        else{
            dp[n][capacity] = knapSackHelper(capacity,val,wt,n-1); 
        }
        return dp[n][capacity];
    }
};

//{ Driver Code Starts.

int main() {
    // Taking total test cases
    int testCases;
    cin >> testCases;
    cin.ignore();
    while (testCases--) {
        // Reading number of items and capacity
        int numberOfItems, capacity;
        vector<int> weights, values;
        string input;
        int number;

        // Read capacity and number of items
        getline(cin, input);
        stringstream ss(input);
        ss >> capacity;      // The first number is the capacity
        ss >> numberOfItems; // The second number is the number of items

        // Read values
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            values.push_back(number);
        }

        // Read weights
        getline(cin, input);
        ss.clear();
        ss.str(input);
        while (ss >> number) {
            weights.push_back(number);
        }

        Solution solution;
        cout << solution.knapSack(capacity, values, weights) << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends